model/riscv_insts_zvkned.sail

```diff
@@ -0,0 +1,550 @@
+/*
+ * Vector Cryptography Extension - NIST Suite: Vecttor AES Block Cipher
+ * ----------------------------------------------------------------------
+ */
+
+/*
+ * Helper functions.
+ * ----------------------------------------------------------------------
+ */
+
+val	 zvk_check_elements : (int, int, int, int) -> bool
+function zvk_check_elements(VLEN, num_elem, LMUL, SEW) = {
+  ((unsigned(vl)%num_elem) != 0) | ((unsigned(vstart)%num_elem) != 0) | (LMUL*VLEN) < (num_elem*SEW)
+}
+
+mapping vv_or_vs : string <-> bits(7) = {
+  "vv" <-> 0b1010001,
+  "vs" <-> 0b1010011,
+}
+
+val	 aes_rotword : bits(32) -> bits(32)
+function aes_rotword(x) = {
+  let a0 : bits (8) = x[ 7.. 0];
+  let a1 : bits (8) = x[15.. 8];
+  let a2 : bits (8) = x[23..16];
+  let a3 : bits (8) = x[31..24];
+  (a0 @ a3 @ a2 @ a1) /* Return Value */
+}
+
+/* VAESEF.[VV, VS] */
+
+mapping vaesef_mnemonic : bits(7) <-> string = {
+  0b1010001 <-> "vaesef.vv",
+  0b1010011 <-> "vaesef.vs",
+}
+
+union clause ast = RISCV_VAESEF : (regidx, regidx, string)
+
+mapping clause encdec = RISCV_VAESEF(vs2, vd, suffix)	       if (haveRVV() & haveZvkned())
+ <-> vv_or_vs(suffix) @ vs2 @ 0b00011 @ 0b010 @ vd @ 0b1110111 if (haveRVV() & haveZvkned())
+
+mapping clause assembly = RISCV_VAESEF(vs2, vd, suffix)
+ <-> vaesef_mnemonic(vv_or_vs(suffix)) ^ spc() ^ vreg_name(vd)
+				       ^ spc() ^ vreg_name(vs2)
+
+function clause execute (RISCV_VAESEF(vs2, vd, suffix)) = {
+  let SEW      = get_sew();
+  let LMUL_pow = get_lmul_pow();
+  let LMUL     = if LMUL_pow < 0 then 0 else LMUL_pow;
+  let VLEN     = int_power(2, get_vlen_pow());
+  let num_elem = get_num_elem(LMUL_pow, SEW);
+
+  if (zvk_check_elements(VLEN, num_elem, LMUL, SEW) == false)
+  then {
+    handle_illegal();
+    RETIRE_FAIL
+  } else {
+    let 'n = num_elem;
+    let 'm = SEW;
+    assert('m == 32);
+
+    let vs2_val : vector('n, dec, bits('m)) = read_vreg(num_elem, SEW, LMUL_pow, vs2);
+    let vd_val  : vector('n, dec, bits('m)) = read_vreg(num_elem, SEW, LMUL_pow, vd);
+    result      : vector('n, dec, bits('m)) = undefined;
+
+    vd_state : bits(128) = undefined;
+    vs2_key  : bits(128) = undefined;
+
```

Comment:
Would it make sense to define both `vd_state` and `vs2_key` within the scope of the `foreach` since they are only used in the loop body and their live range does not exceed one iteration ?

Reply:
ACK., good catch!